kotlinx-coroutines-core/jvm/test/DispatchersToStringTest.kt

```diff
@@ -18,5 +18,10 @@ class DispatchersToStringTest {
         assertEquals("Dispatchers.Default.limitedParallelism(2)", Dispatchers.Default.limitedParallelism(2).toString())
         // Not overridden at all, limited parallelism returns `this`
         assertEquals("DefaultExecutor", (DefaultDelay as CoroutineDispatcher).limitedParallelism(42).toString())
+
+        assertEquals("filesDispatcher", Dispatchers.IO.limitedParallelism(1, "filesDispatcher").toString())
+        assertEquals("json", Dispatchers.Default.limitedParallelism(2, "json").toString())
+        // Not overridden at all, limited parallelism returns `this`
+        assertEquals("DefaultExecutor", (DefaultDelay as CoroutineDispatcher).limitedParallelism(42, "ignored").toString())
```

Comment:
What are the reasons to take the non-orthogonal path here? I think it's much more intuitive to have names that are unconditionally applied.

Reply:
My line of thinking is the following:

* The implementation allows returning `this` which is a nice property to have -- you don't have to care about the underlying/given pool capacity if not necessary
    * Also, mistakes are much more forgiving. E.g. now for cases [like that](https://github.com/mamoe/mirai/blob/dev/mirai-core-utils/src/commonTest/kotlin/net/mamoe/mirai/utils/testFramework/AssertNoCoroutineSuspension.kt#L49) nothing really happens, with the name forcing to return something else it's suddenly more subtle
* I don't really like the idea of being able to give dispatchers names when no new dispatcher is effectively created -- the probability of accidental aliasing just increases with not that many benefits
* Returning `LimitedDispatcher` for the sole purpose of naming is a suboptimal idea -- it slightly changes the observable behaviour and introduces some overhead. It kind of forces us to introduce one more implementation (`NamedDispatcher`?) and maintain it

Reply:
> The implementation allows returning this which is a nice property to have

This can be preserved in the case of `parallelism >= dispatcher.parallelism && newName != null && newName != dispatcher.name`.

> I don't really like the idea of being able to give dispatchers names when no new dispatcher is effectively created -- the probability of accidental aliasing just increases with not that many benefits

Two bureaucratic benefits come to mind immediately:
* Principle of least surprise: specifically, if the user did provide a name for some reason, it's surprising to see it ignored.
* Principle of orthogonality: conceptually, now, the two arguments to `limitedParallelism` are linked together, while they could not be that.

The second concern is practical, I think. Let's look at the issue statement:

> We have specific dispatchers, and in coroutine dumps they all look like `LimitedDispatcher@xxxxxxxx`, and it's not immediately clear where the dispatcher comes from.

I see two possible cases:
1. `"Dispatchers.Default.limitedParallelism(4)"` is already good enough. Then why would we add a new `name` parameter at all?
2. `"Dispatchers.Default.limitedParallelism(4)"` is *not* good enough, and something like `IndexBuilding` is needed instead. In this case, why does the name revert to the not-good-enough `"Dispatchers.Default.limitedParallelism(4)"` option when there are few CPU cores?

As to the accidental aliasing issue: what dangers do you see in it?

> Returning LimitedDispatcher for the sole purpose of naming is a suboptimal idea -- it slightly changes the observable behaviour and introduces some overhead. It kind of forces us to introduce one more implementation (NamedDispatcher?) and maintain it

That's true. Why is this an issue, though? Is an extra call stack frame per `dispatch` a significant problem?

Reply:
Note: discussed it internally and the rationale will be reflected in the public design doc